src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/UserRoleServiceTests.cs

```diff
@@ -372,55 +622,186 @@ await using (var contentDbContext = InMemoryApplicationDbContext(contentDbContex
-
             await using (var userAndRolesDbContext = InMemoryUserAndRolesDbContext(userAndRolesDbContextId))
             await using (var contentDbContext = InMemoryApplicationDbContext(contentDbContextId))
             {
                 var service = SetupUserRoleService(usersAndRolesDbContext: userAndRolesDbContext,
-                    contentDbContext: contentDbContext,
-                    emailTemplateService: emailTemplateService.Object);
-
-                var result = await service.AddReleaseRole(userId, release.Id, Contributor);
+                    contentDbContext: contentDbContext);
 
-                result.AssertRight();
+                var result = await service.AddPublicationRole(userId, publication.Id, Owner);
 
-                emailTemplateService.Verify(mock =>
-                        mock.SendReleaseRoleEmail(user.Email,
-                            It.Is<Release>(p => p.Id == release.Id),
-                            Contributor),
-                    Times.Once);
+                result.AssertBadRequest(UserAlreadyHasResourceRole);
             }
 
             await using (var contentDbContext = InMemoryApplicationDbContext(contentDbContextId))
             {
-                var userReleaseRoles = await contentDbContext.UserReleaseRoles
+                var userPublicationRoles = await contentDbContext.UserPublicationRoles
                     .AsQueryable()
                     .ToListAsync();
-                Assert.Single(userReleaseRoles);
+                Assert.Single(userPublicationRoles);
+
+                Assert.Equal(userPublicationRole.Id, userPublicationRoles[0].Id);
+            }
+        }
+
+        [Fact]
+        public async Task AddPublicationRole_NoUser()
+        {
+            var publication = new Publication();
+
+            var userAndRolesDbContextId = Guid.NewGuid().ToString();
+            var contentDbContextId = Guid.NewGuid().ToString();
+
+            await using (var contentDbContext = InMemoryApplicationDbContext(contentDbContextId))
+            {
+                await contentDbContext.Publications.AddAsync(publication);
+                await contentDbContext.SaveChangesAsync();
+            }
+
+            await using (var userAndRolesDbContext = InMemoryUserAndRolesDbContext(userAndRolesDbContextId))
+            await using (var contentDbContext = InMemoryApplicationDbContext(contentDbContextId))
+            {
+                var service = SetupUserRoleService(usersAndRolesDbContext: userAndRolesDbContext,
+                    contentDbContext: contentDbContext);
+
+                var result = await service.AddPublicationRole(Guid.NewGuid(), publication.Id, Owner);
+
+                result.AssertNotFound();
+            }
+
+            await using (var contentDbContext = InMemoryApplicationDbContext(contentDbContextId))
+            {
+                var userPublicationRoles = await contentDbContext.UserPublicationRoles
+                    .AsQueryable()
```

Comment:
Not sure that you need `.AsQueryable()` here, and can possibly be removed from some other cases throughout this class?

Reply:
Yup, purged them all ta